# Prediciting House Sale Prices With Linear Regression Machine Learning

###### In this project we will work on housing data for the city of Ames, Iowa, USA from 2006 to 2010.

###### The aim of this project is to explore feature engineering (that is how to decide which features will be important in machine learning, and how to manipulate them to function in machine learning

###### More about the data can be found [here](https://doi.org/10.1080/10691898.2011.11889627)

### Loading The Data

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import KFold

pd.options.display.max_columns = 999

In [2]:
housing = pd.read_csv('AmesHousing.tsv', sep='\t')

In [3]:
housing.shape

(2930, 82)

In [4]:
housing.sample(5)

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
74       75  531380080           60        RL           NaN      8880   Pave   
642     643  535303110           20        RL          53.0      8128   Pave   
2018   2019  903233140           45        RM          51.0      6120   Pave   
1772   1773  528366050           20        RL           NaN     12692   Pave   
2739   2740  905451050           20        RL          80.0     12048   Pave   

     Alley Lot Shape Land Contour Utilities Lot Config Land Slope  \
74     NaN       IR1          Lvl    AllPub     Inside        Gtl   
642    NaN       IR1          Lvl    AllPub    CulDSac        Gtl   
2018   NaN       Reg          Lvl    AllPub     Corner        Gtl   
1772   NaN       IR1          Lvl    AllPub     Inside        Gtl   
2739   NaN       Reg          Lvl    AllPub     Inside        Gtl   

     Neighborhood Condition 1 Condition 2 Bldg Type House Style  Overall Qual  \
74        SawyerW        Norm        Norm      1Fam      2Story             7   
642         NAmes        Norm        Norm      1Fam      1Story             6   
2018      BrkSide        Norm        Norm      1Fam      1.5Unf             7   
1772      NoRidge        Norm        Norm      1Fam      1Story             8   
2739      Edwards        Norm        Norm      1Fam      1Story             5   

      Overall Cond  Year Built  Year Remod/Add Roof Style Roof Matl  \
74               5        1994            2002      Gable   CompShg   
642              7        1954            1954        Hip   CompShg   
2018             8        1929            2001      Gable   CompShg   
1772             5        1992            1993        Hip   CompShg   
2739             6        1952            2002      Gable   CompShg   

     Exterior 1st Exterior 2nd Mas Vnr Type  Mas Vnr Area Exter Qual  \
74        VinylSd      VinylSd         None           0.0         Gd   
642       MetalSd      MetalSd      BrkFace          80.0         TA   
2018      Wd Sdng      Wd Sdng         None           0.0         TA   
1772      BrkFace      BrkFace         None           0.0         Gd   
2739      Wd Sdng      Wd Sdng      BrkFace         232.0         TA   

     Exter Cond Foundation Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  \
74           TA      PConc        Gd        TA            No            GLQ   
642          TA     CBlock        TA        TA            No            ALQ   
2018         TA     BrkTil        TA        TA            No            Unf   
1772         TA      PConc        Gd        TA            No            GLQ   
2739         TA       Slab       NaN       NaN           NaN            NaN   

      BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  \
74           695.0            Unf           0.0        253.0          948.0   
642          490.0            Unf           0.0        572.0         1062.0   
2018           0.0            Unf           0.0        832.0          832.0   
1772        1231.0            Unf           0.0       1969.0         3200.0   
2739           0.0            NaN           0.0          0.0            0.0   

     Heating Heating QC Central Air Electrical  1st Flr SF  2nd Flr SF  \
74      GasA         Ex           Y      SBrkr        1222         888   
642     GasA         Gd           Y      SBrkr        1062           0   
2018    GasA         Ex           Y      FuseA         854           0   
1772    GasA         Ex           Y      SBrkr        3228           0   
2739    GasA         Gd           Y      SBrkr        1488           0   

      Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  Bsmt Half Bath  Full Bath  \
74                  0         2110             1.0             0.0          2   
642                 0         1062             1.0             0.0          1   
2018                0          854             0.0             0.0          1   
1772                0         3228          

In [5]:
housing.columns

Index(['Order', 'PID', 'MS SubClass', 'MS Zoning', 'Lot Frontage', 'Lot Area',
       'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities',
       'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1',
       'Condition 2', 'Bldg Type', 'House Style', 'Overall Qual',
       'Overall Cond', 'Year Built', 'Year Remod/Add', 'Roof Style',
       'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type',
       'Mas Vnr Area', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual',
       'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin SF 1',
       'BsmtFin Type 2', 'BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF',
       'Heating', 'Heating QC', 'Central Air', 'Electrical', '1st Flr SF',
       '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Bsmt Full Bath',
       'Bsmt Half Bath', 'Full Bath', 'Half Bath', 'Bedroom AbvGr',
       'Kitchen AbvGr', 'Kitchen Qual', 'TotRms AbvGrd', 'Functional',
       'Fireplaces', 'Fireplace Qu', 'Garage Type', 'Garage Yr Blt',
      

###### We will begin by creating some simple functions that form the basis for our machine learning.

###### We will explore later what we should/could add to these functions.

In [6]:
def transform_features(train):
    return train

In [7]:
def select_features(train):
    return train[['Gr Liv Area','SalePrice']]

In [8]:
def train_and_test(data):
    #split the data between train and test datasets#
    train = data[:1460]
    test = data[1460:]
    #Select only numeric columns#
    train = train.select_dtypes(include = ['integer', 'float'])
    test = test.select_dtypes(include = ['integer', 'float'])
    #Remove our target from the train column#    
    feature_list= train.columns.drop('SalePrice')
    #Create and run the model#
    model = LinearRegression()
    model.fit(train[feature_list], train['SalePrice'])
    predictions = model.predict(test[feature_list])
    #Mean squared error calculation#
    MSE = mean_squared_error(test['SalePrice'], predictions)
    #Relative mean squared error calculation#
    RMSE = np.sqrt(MSE)
    return RMSE

In [9]:
#Quick test on data as is#
transformed_data = transform_features(housing)
filtered_data = select_features(transformed_data)
RMSE = train_and_test(filtered_data)
print("The relative mean square error is {:.2f}".format(RMSE))

The relative mean square error is 57088.25


### Feature Engineering
###### To overcome the missing data problem there are many possible solutions.
 - ###### A)  For this excercise we will keep only columns with <5% missing values
 - ###### B)  Keep text columns with no missing values
 - ###### C)  Keep numerical columns and fill in missing data


###### For now lets start with step A) and drop all columns with 5% missing rows

###### Here is a preview of the missing rows.

In [10]:
housing.isnull().sum()[housing.isnull().sum() >0].sort_values(ascending = False)

Pool QC           2917
Misc Feature      2824
Alley             2732
Fence             2358
Fireplace Qu      1422
Lot Frontage       490
Garage Yr Blt      159
Garage Cond        159
Garage Qual        159
Garage Finish      159
Garage Type        157
Bsmt Exposure       83
BsmtFin Type 2      81
Bsmt Cond           80
Bsmt Qual           80
BsmtFin Type 1      80
Mas Vnr Type        23
Mas Vnr Area        23
Bsmt Half Bath       2
Bsmt Full Bath       2
Garage Cars          1
BsmtFin SF 2         1
BsmtFin SF 1         1
Bsmt Unf SF          1
Total Bsmt SF        1
Garage Area          1
Electrical           1
dtype: int64

In [11]:
print("5% of the missing values represents {} rows".format(len(housing) * 0.05))

5% of the missing values represents 146.5 rows


In [12]:
missing_rows = housing.isnull().sum()
columns_to_remove = missing_rows[missing_rows > (len(housing) *0.05)]
print("The following columns will be removed:","\n")
print(columns_to_remove.sort_values(ascending=False))

The following columns will be removed: 

Pool QC          2917
Misc Feature     2824
Alley            2732
Fence            2358
Fireplace Qu     1422
Lot Frontage      490
Garage Cond       159
Garage Qual       159
Garage Finish     159
Garage Yr Blt     159
Garage Type       157
dtype: int64


In [13]:
housing = housing.drop(columns_to_remove.index, axis=1)

###### Step B) Drop any text columns with 1 or more missing values

In [14]:
missing_text_counts = housing.select_dtypes(include = ['object']).isnull().sum().sort_values(ascending= False)
missing_text_counts.head(8)

Bsmt Exposure     83
BsmtFin Type 2    81
BsmtFin Type 1    80
Bsmt Qual         80
Bsmt Cond         80
Mas Vnr Type      23
Electrical         1
Utilities          0
dtype: int64

In [15]:
#We will only drop the columns above that have missing values#
text_cols_to_drop = missing_text_counts[missing_text_counts>0]
housing = housing.drop(text_cols_to_drop.index, axis=1)

###### Step C) For numerical columns with missing values we shall fill these in. 
###### Many options can be chosen here such as comparing similar rows or inputting average, median or modes.
###### Here we have chosen to inpute the median values (i.e. the most common)


In [16]:
numerical_missing = housing.select_dtypes(include= ['int', 'float']).isnull().sum()
numerical_columns_to_update = numerical_missing[(numerical_missing > 0)].sort_values(ascending=False)
numerical_columns_to_update

Mas Vnr Area      23
Bsmt Half Bath     2
Bsmt Full Bath     2
Garage Area        1
Garage Cars        1
Total Bsmt SF      1
Bsmt Unf SF        1
BsmtFin SF 2       1
BsmtFin SF 1       1
dtype: int64

In [17]:
#Most common values can be calculated using the mode function#
numerical_mode_values = housing[numerical_columns_to_update.index].mode()
numerical_mode_values

Mas Vnr Area  Bsmt Half Bath  Bsmt Full Bath  Garage Area  Garage Cars  \
0           0.0             0.0             0.0          0.0          2.0   

   Total Bsmt SF  Bsmt Unf SF  BsmtFin SF 2  BsmtFin SF 1  
0            0.0          0.0           0.0           0.0

In [18]:
#Convert the above to a replacement dictionary#
replacement_dictionary = numerical_mode_values.to_dict(orient = 'records')[0]
replacement_dictionary

{'Bsmt Full Bath': 0.0,
 'Bsmt Half Bath': 0.0,
 'Bsmt Unf SF': 0.0,
 'BsmtFin SF 1': 0.0,
 'BsmtFin SF 2': 0.0,
 'Garage Area': 0.0,
 'Garage Cars': 2.0,
 'Mas Vnr Area': 0.0,
 'Total Bsmt SF': 0.0}

In [19]:
#Use the fillna() method to replace the missing values#
housing = housing.fillna(replacement_dictionary)

In [20]:
#Check to see if our replacements have worked#
housing.isnull().sum().value_counts()

0    64
dtype: int64

### Feature Engineering (Part 2)
###### Next lets see if we can create more meaningful features

###### We can see that we have columns for the year sold,  year refurbished and the year built. 

###### These are indeed important but also give rise to extra information - the years between buidling and selling, as well as the number of years between building and refurbishment.

In [21]:
years_until_sold = housing['Yr Sold'] - housing['Year Built']
years_until_sold.value_counts().head()

1    218
0    116
2     90
4     76
5     66
dtype: int64

In [22]:
years_until_remod = housing['Yr Sold'] - housing['Year Remod/Add']
years_until_remod.sort_values(ascending= False).head()

230    60
304    60
282    60
314    60
197    60
dtype: int64

###### We should be careful to make sure there are no negative values:

In [23]:
years_until_sold[years_until_sold <0]

2180   -1
dtype: int64

In [24]:
years_until_remod[years_until_remod <0]

1702   -1
2180   -2
2181   -1
dtype: int64

###### We can now create the extra columns and drop these suspicious rows.

In [25]:
housing['Years Before Sale'] = years_until_sold
housing['Years Before Remod'] = years_until_remod

In [26]:
housing = housing.drop([1702,2180,2181], axis=0)
#Can also drop the original year built and remodelled columns#
housing = housing.drop(['Year Built', 'Year Remod/Add'], axis =1 )

### Feature Engineering (Part 3)
###### Next lets drop columns that will not be useful in machine learning or leak data about the final sale

###### The order column is just an index of when the observation was made we could expect no correlation here.

###### The PID is the parcel indentification used with the city website for parcel reviews, this also would be expected to have no correlation

In [27]:
housing.corr()['SalePrice'][['Order', 'PID']].sort_values(ascending= False)

Order   -0.031542
PID     -0.246389
Name: SalePrice, dtype: float64

###### Suprisingly PID does have some correlation although weak.

###### We have lots of colums that would leak information about the final sale price: month sold year sold, sale type, sale condition.

###### For review these columns are fine but in creating a machine learning model we cannot be using information that would be unknown in a prediction

###### So lets drop these columns

In [28]:
housing = housing.drop(["Mo Sold", "Sale Condition", "Sale Type", "Yr Sold"], axis=1)

Now lets update the transform features function to include what we have been performing above.

In [29]:
def transform_features(df):
    #remove numerical columns with les
    missing_data = df.isnull().sum()
    drop_missing_data_columns = missing_data[(missing_data > len(df)*.05)].sort_values()
    df = df.drop(drop_missing_data_columns.index, axis =1)    
    #Keep text columns with no missing values#
    text_missing = df.select_dtypes(include = ['object']).isnull().sum().sort_values(ascending = False)
    text_missing_cols = text_missing[text_missing >0]
    df = df.drop(text_missing_cols.index, axis =1)  
    #Keep remaining numerical columns and fill in missing data#
    numerical_missing = df.select_dtypes(include = ['int','float']).isnull().sum()
    fixable_numerical_columns = numerical_missing[numerical_missing > 0].sort_values()
    #Replace missing numerics with the mode of each respective column#
    value_replacement_dictionary = df[fixable_numerical_columns.index].mode().to_dict(orient='records')[0]
    df = df.fillna(value_replacement_dictionary)
    #Create Years Before Sale and Years Since Remod#
    years_sold = df['Yr Sold'] - df['Year Built']
    years_since_remod = df['Yr Sold'] - df['Year Remod/Add']
    df['Years Before Sale'] = years_sold
    df['Years Since Remod'] = years_since_remod
    #Remove rows with clearly incorrect inputs (negative year differences) #
    df = df.drop([1702, 2180, 2181], axis=0)
    #Remove columns with no relevance or leak target information#
    df = df.drop(["PID", "Order", "Mo Sold", "Sale Condition", "Sale Type", "Year Built", "Year Remod/Add"], axis=1)
    return df

#Test the updated model#
df = pd.read_csv("AmesHousing.tsv", delimiter="\t")
transform_df = transform_features(df)
filtered_df = select_features(transform_df)
rmse = train_and_test(filtered_df)

rmse

55275.36731241307

###### Our Root Mean Squared Error has reduced from 57088 to 55275 (a 3% reduction)

### Feature Selection

###### What features would be good for our linear regression model?

###### Clearly ones that have good linear correlation, so lets see what numerical columns correlate well:

In [30]:
numerical_housing = transform_df.select_dtypes(include = ['int', 'float'])
absolute_num_correlation = numerical_housing.corr()['SalePrice'].abs().sort_values(ascending= False)
absolute_num_correlation

SalePrice            1.000000
Overall Qual         0.801206
Gr Liv Area          0.717596
Garage Cars          0.648361
Total Bsmt SF        0.644012
Garage Area          0.641425
1st Flr SF           0.635185
Years Before Sale    0.558979
Full Bath            0.546118
Years Since Remod    0.534985
Mas Vnr Area         0.506983
TotRms AbvGrd        0.498574
Fireplaces           0.474831
BsmtFin SF 1         0.439284
Wood Deck SF         0.328183
Open Porch SF        0.316262
Half Bath            0.284871
Bsmt Full Bath       0.276258
2nd Flr SF           0.269601
Lot Area             0.267520
Bsmt Unf SF          0.182751
Bedroom AbvGr        0.143916
Enclosed Porch       0.128685
Kitchen AbvGr        0.119760
Screen Porch         0.112280
Overall Cond         0.101540
MS SubClass          0.085128
Pool Area            0.068438
Low Qual Fin SF      0.037629
Bsmt Half Bath       0.035875
3Ssn Porch           0.032268
Yr Sold              0.030358
Misc Val             0.019273
BsmtFin SF

###### Clearly there is a trade off here, poor correlating features can give unwanted noise (and reduced accuracy) as well as increasing the model creation time.

###### To few and we may have suffer bias.

###### For now lets select any feature column with a value of 0.4 or greater correlation.

In [31]:
corr_features_to_remove = absolute_num_correlation[absolute_num_correlation <0.4]
corr_features_to_remove

Wood Deck SF       0.328183
Open Porch SF      0.316262
Half Bath          0.284871
Bsmt Full Bath     0.276258
2nd Flr SF         0.269601
Lot Area           0.267520
Bsmt Unf SF        0.182751
Bedroom AbvGr      0.143916
Enclosed Porch     0.128685
Kitchen AbvGr      0.119760
Screen Porch       0.112280
Overall Cond       0.101540
MS SubClass        0.085128
Pool Area          0.068438
Low Qual Fin SF    0.037629
Bsmt Half Bath     0.035875
3Ssn Porch         0.032268
Yr Sold            0.030358
Misc Val           0.019273
BsmtFin SF 2       0.006127
Name: SalePrice, dtype: float64

In [32]:
transform_df = transform_df.drop(corr_features_to_remove.index, axis=1)

In [33]:
len(transform_df.columns.tolist())

39

###### We still have 38 features (columns) for the model (the 39th being our target SalePrice) 

###### From the data source literature ([here](http://jse.amstat.org/v19n3/decock/DataDocumentation.txt)) we can see multiple columns should be categorical

In [34]:
#Here is a list of columns that according to the documentation are meant to be categorical#
nominal_features = ["PID", "MS SubClass", "MS Zoning", "Street", "Alley", "Land Contour", "Lot Config", "Neighborhood", 
                    "Condition 1", "Condition 2", "Bldg Type", "House Style", "Roof Style", "Roof Matl", "Exterior 1st", 
                    "Exterior 2nd", "Mas Vnr Type", "Foundation", "Heating", "Central Air", "Garage Type", 
                    "Misc Feature", "Sale Type", "Sale Condition"]

###### - Which columns are currently numerical but should be classed as categorical (as the numbers dont have correct scalable meaning

###### - If a categorical column has vast unique values ( >10 say) then we will generate large amounts of columns (as each category variable will get a category with data of 0 or 1)

In [35]:
#Out of the remaining feautres which should be categorical#
transform_cat_cols = []
for col in nominal_features:
    if col in transform_df.columns:
        transform_cat_cols.append(col)
print('There are', len(transform_cat_cols), 'columns that are interpretted as numeric but are really categorical:')
print('\n')
print(transform_cat_cols)

There are 16 columns that are interpretted as numeric but are really categorical:


['MS Zoning', 'Street', 'Land Contour', 'Lot Config', 'Neighborhood', 'Condition 1', 'Condition 2', 'Bldg Type', 'House Style', 'Roof Style', 'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Foundation', 'Heating', 'Central Air']


In [36]:
#How many unique values are there in each category#
unique_values = transform_df[transform_cat_cols].apply(lambda x: len(x.value_counts())).sort_values(ascending=False)
unique_values

Neighborhood    28
Exterior 2nd    17
Exterior 1st    16
Condition 1      9
Roof Matl        8
House Style      8
Condition 2      8
MS Zoning        7
Heating          6
Foundation       6
Roof Style       6
Bldg Type        5
Lot Config       5
Land Contour     4
Central Air      2
Street           2
dtype: int64

###### We have a range of options here and have to make an arbitary cut off.

###### Lets use a cutoff of 10 features, this will remove only 3 of the above features

In [37]:
columns_to_drop_10_plus = unique_values[unique_values >10].index
transform_df = transform_df.drop(columns_to_drop_10_plus, axis=1)
transform_df.sample(5)

MS Zoning Street Lot Shape Land Contour Utilities Lot Config Land Slope  \
567         FV   Pave       Reg          Lvl    AllPub     Inside        Gtl   
2477        RL   Pave       IR1          Lvl    AllPub     Inside        Gtl   
2379        RL   Pave       IR1          Lvl    AllPub     Inside        Gtl   
617         RL   Pave       Reg          Lvl    AllPub     Inside        Gtl   
1226        RL   Pave       Reg          Lvl    AllPub     Inside        Gtl   

     Condition 1 Condition 2 Bldg Type House Style  Overall Qual Roof Style  \
567         Norm        Norm    TwnhsE      2Story             6      Gable   
2477        Norm        Norm      1Fam      2Story             6      Gable   
2379        Norm        Norm      1Fam      2Story             9        Hip   
617        Feedr        Norm      1Fam      1Story             6      Gable   
1226        Norm        Norm      1Fam        SLvl             6        Hip   

     Roof Matl  Mas Vnr Area Exter Qual Exter Cond Foundation  BsmtFin SF 1  \
567    CompShg          76.0         Gd         TA      PConc         294.0   
2477   CompShg           0.0         TA         TA      PConc           0.0   
2379   CompShg         215.0         Gd         TA      PConc        1369.0   
617    CompShg           0.0         Gd         TA     CBlock        1173.0   
1226   CompShg           0.0         TA         TA     CBlock         831.0   

      Total Bsmt SF Heating Heating QC Central Air  1st Flr SF  Gr Liv Area  \
567           612.0    GasA         Ex           Y         612         1224   
2477          941.0    GasA         Ex           Y         941         1837   
2379         1738.0    GasA         Gd           Y        1738         2589   
617          1680.0    GasA         TA           Y        1691         1691   
1226          992.0    GasA         Gd           Y        1661         1661   

      Full Bath Kitchen Qual  TotRms AbvGrd Functional  Fireplaces  \
567           2           Gd              4        Typ           0   
2477          2           TA              7        Typ           0   
2379          2           Ex             11        Typ           1   
617           1           TA              5        Typ           0   
1226          1           Gd              8        Typ           1   

      Garage Cars  Garage Area Paved Drive  SalePrice  Years Before Sale  \
567           2.0        528.0           Y     173733                  0   
2477          2.0        688.0           Y     187000                  9   
2379          3.0        831.0           Y     412083                  0   
617           2.0        550.0           Y     175000                 42   
1226          1.0        377.0           Y     165500                 53   

      Years Since Remod  
567                   0  
2477                  8  
2379                  0  
617                  42  
1226                 12

###### Next lets remove categories that have are ordered but cannot be easily converted to numbers

###### For instance Kitchen Quality is:
 - ###### Ex = 	Excellent
 - ###### Gd =	Good
 - ###### TA =	Typical/Average
 - ###### Fa =	Fair
 - ###### Po =	Poor
 
######  This is very difficult to assign weightings for in machine learning, is average a 0.5 out of 1, or is excellent orders of magnitude greater than average?
 
###### With correct domain knowledge this could possibly be known, but here we will simply have to remove these columns.

In [38]:
categories_to_remove = list(set(transform_df.select_dtypes(include = 'object').columns) 
     - set(unique_values[unique_values <10].index))
categories_to_remove

['Kitchen Qual',
 'Functional',
 'Utilities',
 'Exter Qual',
 'Lot Shape',
 'Paved Drive',
 'Exter Cond',
 'Land Slope',
 'Heating QC']

In [39]:
transform_df = transform_df.drop(categories_to_remove, axis =1)

In [40]:
transform_df.sample(5)

MS Zoning Street Land Contour Lot Config Condition 1 Condition 2  \
2858        RL   Pave          Lvl     Inside        Norm        Norm   
860         RL   Pave          Lvl     Inside        Norm        Norm   
1139        RL   Pave          Lvl     Inside        Norm        Norm   
1127        FV   Pave          Lvl     Inside        Norm        Norm   
2267        RL   Pave          Lvl     Inside        Norm        Norm   

     Bldg Type House Style  Overall Qual Roof Style Roof Matl  Mas Vnr Area  \
2858      1Fam      2Story             6      Gable   CompShg           0.0   
860       1Fam      SFoyer             5      Gable   CompShg           0.0   
1139      1Fam      2Story             6      Gable   CompShg          38.0   
1127      1Fam      2Story             7      Gable   CompShg           0.0   
2267      1Fam      1Story             8      Gable   CompShg          86.0   

     Foundation  BsmtFin SF 1  Total Bsmt SF Heating Central Air  1st Flr SF  \
2858      PConc           0.0          817.0    GasA           Y         940   
860      CBlock         660.0          768.0    GasA           Y         768   
1139      PConc         362.0          754.0    GasA           Y         754   
1127      PConc           0.0          813.0    GasA           Y         822   
2267      PConc         250.0         1267.0    GasA           Y        1312   

      Gr Liv Area  Full Bath  TotRms AbvGrd  Fireplaces  Garage Cars  \
2858         1550          1              7           1          1.0   
860           768          1              5           0          1.0   
1139         1609          2              6           0          2.0   
1127         1665          2              7           0          2.0   
2267         1312          2              5           0          2.0   

      Garage Area  SalePrice  Years Before Sale  Years Since Remod  
2858        318.0     195000                 86                 56  
860         396.0     133900                 37                  6  
1139        525.0     182000                 13                 13  
1127        562.0     205950                  1                  1  
2267        471.0     180500                  6                  6

In [41]:
#The pd.get_dummies() method creates columns based on values in a column#

#For instance a column of eye colours could be blue, red, green#

#pd.get_dummies() will create 3 columns and for each row assign a 1 (true)#

# In each of the columns that was the true eye colour#

#Here is a quick preview of how that works on the Lot Config column#

pd.get_dummies(transform_df['Lot Config']).head()

Corner  CulDSac  FR2  FR3  Inside
0       1        0    0    0       0
1       0        0    0    0       1
2       1        0    0    0       0
3       1        0    0    0       0
4       0        0    0    0       1

###### The method returns a dataframe of the columns to which the operation was applied

###### This just means we will need to join this to our original dataframe and then remove the 'pre-dummy' category columns

In [42]:
transform_df = pd.concat([transform_df, pd.get_dummies(transform_df.select_dtypes(include=['object']))])
transform_df = transform_df.drop(transform_df.select_dtypes(include=['object']).columns.tolist(), axis=1)

###### With these operations done we should now only have numerical columns

In [43]:
transform_df.dtypes.value_counts()

float64    90
dtype: int64

###### Excellent! 90 columns (variables) all of which are numerical

###### With this success lets update the select features function

In [44]:
#Add in the correlation coefficient (how well a feature correlates with price)#
#And unique_threshold (how many dummy variables to allow)#
#Into the method parameters to allow easier experimentation of best values#
def select_features(df, coeff_threshold=0.4, unique_threshold=10):
    #Create list of columns with category order but cannot convert to numeric#
    ordered_categories = ['Land Slope', 'Paved Drive', 'Utilities', 'Functional', 'Exter Cond',
                            'Lot Shape', 'Exter Qual', 'Heating QC', 'Kitchen Qual']
    #Remove these columns#
    df = df.drop(ordered_categories, axis=1)
    numerical_df = df.select_dtypes(include = ['int', 'float'])
    absolute_correlation_coeff = numerical_df.corr()['SalePrice'].abs()
    df = df.drop(absolute_correlation_coeff[absolute_correlation_coeff < coeff_threshold].index, axis=1)
    #Set the nominal features inside the method so it is self contained#
    nominal_features = ["PID", "MS SubClass", "MS Zoning", "Street", "Alley", "Land Contour", "Lot Config", "Neighborhood", 
                    "Condition 1", "Condition 2", "Bldg Type", "House Style", "Roof Style", "Roof Matl", "Exterior 1st", 
                    "Exterior 2nd", "Mas Vnr Type", "Foundation", "Heating", "Central Air", "Garage Type", 
                    "Misc Feature", "Sale Type", "Sale Condition"]
    #Check to see if these exist in the dataframe#
    transform_cat_cols = []
    for col in nominal_features:
        if col in df.columns:
            transform_cat_cols.append(col)
    #Calculate the number of unique values in each of these columns in the dataframe#
    uniqueness_count = df[transform_cat_cols].apply(lambda col: len(col.value_counts()))
    #Find category columns with more unique values than the unique threshold#
    drop_unique_col = uniqueness_count[uniqueness_count > unique_threshold].index
    #Then drop these columns#
    df = df.drop(drop_unique_col, axis =1)
    #With remaining category (object columns) convert each unique value to dummy columns#
    text_columns = df.select_dtypes(include = ['object'])
    df = pd.concat([df, pd.get_dummies(text_columns)], axis=1)
    #Remove the columns used to create the dummy columns#
    df = df.drop(text_columns, axis=1)
    return df

In [45]:
df = pd.read_csv("AmesHousing.tsv", delimiter="\t")
transform_df = transform_features(df)
filtered_df = select_features(transform_df)
rmse = train_and_test(filtered_df)

print("The relative mean square error is {:.2f}".format(rmse))

The relative mean square error is 35064.65


###### Our Root Mean Squared Error has reduced from 57088 to 55275 and then to 35066 (a 39% total reduction)

### Cross Validation

###### Having performed the transformation steps the next thing on the list is to perform the cross validation.

###### This has the effect of shuffling the locations of the train / testing splits

###### We can do this to work out the best representation of our data i.e. best homogenity for the best model

###### So lets explore how this would work in our train and test function

In [46]:
#Here is our original train and test function. This uses 'holdout' validation
def train_and_test(data):
    #split the data between train and test datasets#
    train = data[:1460]
    test = data[1460:]
    #Select only numeric columns#
    train = train.select_dtypes(include = ['integer', 'float'])
    test = test.select_dtypes(include = ['integer', 'float'])
    #Remove our target from the train column#    
    feature_list= train.columns.drop('SalePrice')
    #Create and run the model#
    model = LinearRegression()
    model.fit(train[feature_list], train['SalePrice'])
    predictions = model.predict(test[feature_list])
    #Mean squared error calculation#
    MSE = mean_squared_error(test['SalePrice'], predictions)
    #Relative mean squared error calculation#
    RMSE = np.sqrt(MSE)
    return RMSE

###### In the above function we have implemented holdout validation where the data is split into 2 portions, this is performed only once.

###### In K-Fold Cross Validation this can be performed multiple times

###### [Click here for a quick example on a K-fold cross validation using k = 5](https://scikit-learn.org/stable/_images/grid_search_cross_validation.png)

###### Here we can see the model predictions are run 5 times with the location of the testing data altered each time

###### In all K-Fold cross validation the data is randomised before it processed.

In [47]:
def train_and_test(df, k=0):
    #Select only numeric columns and Remove the target#
    numerical_df = df.select_dtypes(include = ['integer', 'float'])
    feature_list= numerical_df.columns.drop('SalePrice')
    model = LinearRegression()
    if k == 0:
        #Perform holdout validation#
        #Data split to train and test#
        train = df[:1460]
        test = df[1460:]
        #Create and test model#
        model.fit(train[feature_list], train['SalePrice'])
        predictions = model.predict(test[feature_list])
        mse = mean_squared_error(test['SalePrice'], predictions)
        rmse = np.sqrt(mse)
        print("K = {} selected. Relative Mean Standard error was {}".format(k, rmse))    

    if k == 1:
        #Perform simple cross validation with randomisation of the data first#
        #Sample performs a random selection of data, frac =1 sets this to all data#
        shuffled_df = df.sample(frac=1,)
        #Create and test model#
        train = df[:1460]
        test = df[1460:]
        model.fit(train[feature_list], train['SalePrice'])
        predictions_one = model.predict(test[feature_list])
        mse_one = mean_squared_error(test['SalePrice'], predictions_one)
        rmse_one = np.sqrt(mse_one)
        #Reverse train and test and repeat#
        train = df[1460:]
        test = df[:1460]
        model.fit(train[feature_list], train['SalePrice'])
        predictions_two = model.predict(test[feature_list])
        mse_two = mean_squared_error(test['SalePrice'], predictions_two)
        rmse_two = np.sqrt(mse_two)
        rmses = [rmse_one, rmse_two]
        average_rmse = np.mean([rmse_one, rmse_two])
        print("\n")
        print("K = {} selected. Relative Mean Standard errors were: {} with an average of {}".format(k, rmses, average_rmse))    
        
    if k >= 2:
        #Use SKLearns KFold method. Shuffle randomises data before splitting into batches#
        kf = KFold(n_splits = k, shuffle = True)  
        rmses = []
        #KFold gives index ranges to each of the batches#
        for train_index, test_index in kf.split(df):
            train = df.iloc[train_index]
            test = df.iloc[test_index]
            model.fit(train[feature_list], train['SalePrice'])
            predictions = model.predict(test[feature_list])
            mse = mean_squared_error(test['SalePrice'], predictions)
            rmse = np.sqrt(mse)
            rmses.append(rmse)
            average_rmse = np.mean(rmses)
        print("\n")
        print("K = {} selected. Relative Mean Standard errors were: {} with an average of {}".format(k, rmses, average_rmse))            

###### Next we shall test the model using our various K parameters

In [49]:
df = pd.read_csv("AmesHousing.tsv", delimiter="\t")
for i in range(7):
    transform_df = transform_features(df)
    filtered_df = select_features(transform_df)
    rmse = train_and_test(filtered_df, k=i)

    rmse

K = 0 selected. Relative Mean Standard error was 35064.65061930173


K = 1 selected. Relative Mean Standard errors were: [35064.65061930173, 29641.240536401056] with an average of 32352.94557785139


K = 2 selected. Relative Mean Standard errors were: [34168.92644378316, 30037.25469525408] with an average of 32103.09056951862


K = 3 selected. Relative Mean Standard errors were: [28188.950583605576, 35807.07114283821, 31063.952412679755] with an average of 31686.658046374516


K = 4 selected. Relative Mean Standard errors were: [29112.05238507154, 37858.633609553144, 28863.074008142143, 29762.187482750855] with an average of 31398.98687137942


K = 5 selected. Relative Mean Standard errors were: [33142.4409730235, 26203.68250498017, 38955.44483177554, 29228.791519260354, 29466.719600026772] with an average of 31399.415885813265


K = 6 selected. Relative Mean Standard errors were: [30014.56240800464, 28612.25057158538, 30221.531399468462, 42159.03021891922, 28150.734639352708, 29070.55

###### The lowest mean standard error seen is with K=5 (5 folds)

###### Our Root Mean Squared Error has reduced from 57088 to 31399 (a 45% total reduction)